0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+This prevents `Store` from being a singleton, and in this case, means that there's more special-case handling needed.
+
+In this RFC, I went with keeping the parameters inherent to `Store` rather than attempting to thread through a `params` field into `CircuitInstruction`.
+I'm worried that trying to do too much there could interact very oddly with the `.clbits` field, and would lead to an odd split where some things were parametrised in `CircuitInstruction` and others not.
+Until it's clear how _all_ instructions should be parametrised, I elected to keep the new object in the same sort of special case as the other classical operations (the control-flow operations).
+
+
+### What does the API for `QuantumCircuit.compose` expand to?
+
+`QuantumCirucit.compose` is in effect an inlining operation.
+It's not clear to me whether it should be possible for a circuit to involve variables whose names conflict in scope with the program it's being inlined into, or what if any automatic variable-renaming methods we should make available.
+
+If there are no conflicts in the variable names, it seems expected to me that the lists of input variables would catenate, and so on.
+Similarly, two declared `Var` instances that conflict (either by having incompatible initialisers, or one being an input, one not, etc) cannot be allowed to coexist, but does this mean that the composition should be rejected, or should we implement a renaming?
+
+
+### Is it necessary to add an `expr.Index` immediately?
+
+At the moment, it's inconvenient to get a measurement result from a `Clbit` into one bit of a declared classical variable; it effectively needs to be done as
+```python
+with qc.if_test(bit_0) as else_:
+    qc.store(my_var, expr.bit_or(my_var, 0b0001))
+with else_:
+    qc.store(my_var, expr.bit_and(my_var, 0b1110))
+
+with qc.if_test(bit_1) as else_:
+    qc.store(my_var, expr.bit_or(my_var, 0b0010))
+with else_:
+    qc.store(my_var, expr.bit_and(my_var, 0b1101))
+
+# ... and so on.
+```
+
+This is not exactly efficient, nor convenient.
+
+Much of this can probably be avoided by judicious use of `ClassicalRegister` instead stores into other variables, but it's worth bringing up as a question now.
+`ClassicalRegister` would not be fully suitable when only part of an object needs filling with the results from bits.
+
+In general, one can imagine having an `expr.Index` expression that produces bit-typed l-values, and allowing stores into that.
+Currently this RFC does not propose this to avoid scope creep, but I don't know if it's something that ought to be done now.
+
+
+## Future Extensions
+
+### Add support for circuit outputs
+
+OpenQASM 3 supports an `output` specifier on variables for rich returns from programs.
+Qiskit currently doesn't have a way of representing this through its result types, and it's not clear when hardware vendors will have support for this.
+For ease of user expectations, it's left til later.
```

Comment:
Maybe I'm bending things too much. I feel the current approach is that every classical wire has an implicit `output` right now.

Reply:
They do, and the OQ3 spec even defines that behaviour to make it consistent-ish with OQ2-era Qiskit behaviour.  In practice, I shied away from defining this because I think the harder questions to answer are hardware-vendor side.

Reply:
This actually touches on my primary question about this RFC, which was that I was expecting it to simultaneously address circuit outputs. At the very least, I can imagine wanting to NOT return some values that I used as intermediaries in my circuit.

Reply:
I can certainly imagine that too.  My intent with this RFC wasn't to imply that all variables will always be `output` for sure.  I'm fine if we'd like to discuss including `output` in this RFC directly (especially now it's pushed to 1.0) - my initial concern was that I didn't want to specify an interface for something that (at least) IBM hardware definitely has no way of representing in the timescales that Qiskit could implement it.

I think it's relatively easy to imagine ways to designate `output` variables in this form. Some (non-exclusive) options are
- have an `output=True` keyword on `add_var` (but not `add_capture` or `add_input`)
- have an `add_output` function with the same signature as `add_var`
- have an `outputs=...` field in the `QuantumCircuit` constructor

The bulk of the remaining `output` work will then be in the results interfaces of the primitives and `backend.run`, and I think the design process might be smoother and more focussed for other affected teams if we separate that out.

Reply:
All of those options seem reasonable.